### PR TITLE
[sparse] add bcoo_sort_indices

### PR DESCRIPTION
**Pull request by jakevdp**

### CODE REVIEW

**Review by froystig — APPROVED** [review]

File: jax/experimental/sparse/bcoo.py [hunk 1]
```diff
@@ -119,6 +119,22 @@ def _bcoo_sum_duplicates_unbatched(data, indices, *, shape, nse, remove_zeros):
...
   return data_unique, indices_unique, nse
 
+def _bcoo_sort_indices(data, indices, shape):
+  props = _validate_bcoo(data, indices, shape)
+  def f(data, indices):
+    _, N = indices.shape
+    if N == 0:
+      return data, indices
+    elif data.ndim > 1:
+      i = jnp.lexsort(indices[:, ::-1].T)
```
Comment (soraros) [inline jax/experimental/sparse/bcoo.py:129]: Will unfolding `lexsort` here be helpful since that way we can save a `gather` in `indices[i]`.

Comment (jakevdp) [reply jax/experimental/sparse/bcoo.py:129]: Im not sure what you mean by unfolding

Comment (soraros) [reply jax/experimental/sparse/bcoo.py:129]: *inline

Comment (jakevdp) [reply jax/experimental/sparse/bcoo.py:129]: Thanks, but I'm not exactly sure how that applies here. Could you give an example of what you have in mind?

Comment (soraros) [reply jax/experimental/sparse/bcoo.py:129]: I mean inside `lexsort` the sorted indices are computed, and only to be thrown away in its return statement `lax.sort((*keys[::-1], iota), dimension=axis, num_keys=len(keys))[-1]`. So I've been thinking maybe if we inline `lexsort` here we can save a indexing which is linear wrt. `N`.

Comment (jakevdp) [reply jax/experimental/sparse/bcoo.py]: Thanks, Done.
